### ISSUE TITLE
Trouble loading CPython class in Java: undefined symbol: Py_IsInitialized

### ISSUE THREAD

**Issue by KeithCaceres** [LeeKamentsky/python-javabridge#102]
I'm sure this must be something I'm doing wrong, because I'm pretty new to Python, but so far I haven't been able to figure it out. Hoping you can help.

I'm trying to load and use the CPython class from an Oracle Java 7 program. It's able to find cpython.jar and the native libjava2cpython.so library just fine, but it doesn't seem able to find the Python 2.7 libraries to link against. I get this:

```
$ java -cp .:jars/* Test
java: symbol lookup error: /home/keith/spark/SparkAMPSAPI/lib/libjava2cpython.so: undefined symbol: Py_IsInitialized
```

I have boiled down the problem to this simple Java program:

```
$ cat Test.java 
import org.cellprofiler.javabridge.CPython;

public class Test {
	public static void main(String[] args) {
		CPython cpy = new CPython();
	}
}
```

I'm using an Ubuntu 16.04 LTS 64-bit virtual machine that seems to have Python 2.7 and Python 3.5 installed (wondering if that is the problem), though Python 2.7 is selected via alternatives.

I have tried various settings for LD_LIBRARY_PATH (or alternatively java.library.path). At a minimum I have it set to point to the directory containing these:

```
$ ls -l lib/
total 13728
-rwxrwxr-x 1 keith keith  1505000 Dec  6 16:12 _javabridge.so
-rwxrwxr-x 1 keith keith    58528 Dec  6 16:12 libjava2cpython.so
```

I have also tried adding various system paths to it that contain the python installation libs such as libpython2.7.so.

In addition I tried placing the system's libpython2.7.so in LD_PRELOAD, but this caused a segmentation fault...

```
$ export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libpython2.7.so
$ java -cp .:jars/* Test #
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f1f01cea7f4, pid=5639, tid=139771157948160
#
# JRE version: Java(TM) SE Runtime Environment (7.0_80-b15) (build 1.7.0_80-b15)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.80-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libpython2.7.so+0xf67f4]  PyString_AsString+0x4
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/keith/spark/SparkAMPSAPI/hs_err_pid5639.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
Aborted (core dumped)
```

I then set the ulimit to allow a core file and attempted to debug the jvm via gdb, but gdb dumped core too!

NOTE: I have no problem if I do the following from the Python interpreter though...

```
$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> import javabridge
>>> 
>>> javabridge.start_vm()
>>> cpython = javabridge.JClassWrapper('org.cellprofiler.javabridge.CPython')()
>>> d = javabridge.JClassWrapper('java.util.Hashtable')()
>>> result = javabridge.JClassWrapper('java.util.ArrayList')()
>>> d.put("result", result)
>>> 
>>> cpython.execute(
...     'import javabridge\n'
...     'x = { "foo":"bar"}\n'
...     'ref_id = javabridge.create_and_lock_jref(x)\n'
...     'javabridge.JWrapper(result).add(ref_id)', d, d)
>>> 
>>> cpython.execute(
...     'import javabridge\n'
...     'ref_id = javabridge.to_string(javabridge.JWrapper(result).get(0))\n'
...     'assert javabridge.redeem_jref(ref_id)["foo"] == "bar"\n'
...     'javabridge.unlock_jref(ref_id)', d, d)
>>> 
>>> 
```

But I need to use CPython from a Java program.

Thanks in advance for any assistance!

-Keith

**Comment by LeeKamentsky** [comment]
I also am running Ubuntu 16.04 right now. I've verified the problem and will take a look at it.

**Comment by LeeKamentsky** [comment]
The problem seems to be the code that looks up the location of libpython2.7.so - it expects to find it in /usr/lib, but on my system, it's in /usr/lib/x86_64-linux-gnu/libpython2.7.so. A workaround is this:

    System.setProperty("python.location", "/usr/lib/x86_64-linux-gnu/libpython2.7.so");
   org.cellprofiler.javabridge.CPython().exec("print \"Hello, world.\"\n");

**Comment by LeeKamentsky** [comment]
@KeithCaceres it should be fixed now, please try it out and feel free to reopen if it doesn't work. Thanks for reporting this.

**Comment by KeithCaceres** [comment]
Hi Lee,

Thanks much for the quick turnaround time on this! I downloaded the repo with your fix as a zip and attempted to build it locally. Originally I had downloaded javabridge-1.0.14.tar.gz and built it using "python setup.py build". This worked fine after I installed the NumPy package.

But on this new zip with your fix, I'm getting a strange error during the build:

```
keith@smithers:~/python-javabridge-master$ python setup.py build
running build
running build_py
creating build
creating build/lib.linux-x86_64-2.7
creating build/lib.linux-x86_64-2.7/javabridge
copying javabridge/jutil.py -> build/lib.linux-x86_64-2.7/javabridge
copying javabridge/__init__.py -> build/lib.linux-x86_64-2.7/javabridge
copying javabridge/wrappers.py -> build/lib.linux-x86_64-2.7/javabridge
copying javabridge/noseplugin.py -> build/lib.linux-x86_64-2.7/javabridge
copying javabridge/locate.py -> build/lib.linux-x86_64-2.7/javabridge
creating build/lib.linux-x86_64-2.7/javabridge/tests
copying javabridge/tests/test_javabridge.py -> build/lib.linux-x86_64-2.7/javabridge/tests
copying javabridge/tests/test_cpython.py -> build/lib.linux-x86_64-2.7/javabridge/tests
copying javabridge/tests/__init__.py -> build/lib.linux-x86_64-2.7/javabridge/tests
copying javabridge/tests/test_jutil.py -> build/lib.linux-x86_64-2.7/javabridge/tests
copying javabridge/tests/test_wrappers.py -> build/lib.linux-x86_64-2.7/javabridge/tests
creating build/lib.linux-x86_64-2.7/javabridge/jars
copying javabridge/jars/rhino-1.7R4.jar -> build/lib.linux-x86_64-2.7/javabridge/jars
running build_ext
javac -source 6 -target 6 java/org/cellprofiler/runnablequeue/RunnableQueue.java
warning: [options] bootstrap class path not set in conjunction with -source 1.6
1 warning
jar cf build/lib.linux-x86_64-2.7/javabridge/jars/runnablequeue.jar -C java org/cellprofiler/runnablequeue/RunnableQueue$1.class -C java org/cellprofiler/runnablequeue/RunnableQueue.class
javac -source 6 -target 6 java/org/cellprofiler/javabridge/test/RealRect.java
warning: [options] bootstrap class path not set in conjunction with -source 1.6
1 warning
jar cf build/lib.linux-x86_64-2.7/javabridge/jars/test.jar -C java org/cellprofiler/javabridge/test/RealRect.class
javac -source 6 -target 6 java/org/cellprofiler/javabridge/CPython.java java/org/cellprofiler/javabridge/CPythonInvocationHandler.java
warning: [options] bootstrap class path not set in conjunction with -source 1.6
Note: java/org/cellprofiler/javabridge/CPythonInvocationHandler.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 warning
jar cf build/lib.linux-x86_64-2.7/javabridge/jars/cpython.jar -C java org/cellprofiler/javabridge/CPython.class -C java org/cellprofiler/javabridge/CPython$StackFrame.class -C java org/cellprofiler/javabridge/CPython$WrappedException.class -C java org/cellprofiler/javabridge/CPythonInvocationHandler.class -C java org/cellprofiler/javabridge/CPythonInvocationHandler.class
error: [Errno 2] No such file or directory
```

It looks like the slash between "java" and "org" is somehow getting changed to a space in paths like:

java org/cellprofiler/javabridge/CPythonInvocationHandler.class

I looked in setup.py and the paths look correct. Any idea what might be causing this?

Thanks,

Keith

**Comment by KeithCaceres** [comment]
Hi Lee,

I patched my javabridge-1.0.14 installation directory with your change to java/org_cellprofiler_javabridge_CPython.c and rebuilt it. I can confirm that your fix worked fine.

I changed my Test.java to use this example from the docs:

http://pythonhosted.org/javabridge/java2python.html

Just wanted to note that the example needs some repairs to the python script in order to work:

```
        script.append("import os\n");
        script.append("import javabridge\n");
        script.append("root = javabridge.to_string(root)\n");    // Added \n to string literal
        script.append("result = javabridge.JWrapper(result)\n");    // Added \n to string literal
        script.append("for path, dirnames, filenames in os.walk(root):\n");
        script.append("  if 'waldo' in filenames:\n");
        script.append("     result.add(path)\n");
```

Thanks!

Keith

**Comment by KeithCaceres** [comment]
No worries on the build issue. I had missed doing this first:

cython *.pyx
